psi4/driver/p4util/solvers.py

```diff
@@ -310,3 +310,159 @@ def extrapolate(self, out=None):
             out.axpy(c, self.state[num])
 
         return out
+
+
+# Generalized Davidson solver
+
+def davidson_solver(ax_function, preconditioner, guess, e_conv=1.0E-8,
+        r_conv=None, no_eigs=1, max_vecs_per_root=20, maxiter=100):
+    """
+    Solves for the lowest few eigenvalues and eigenvectors of the given real symmetric matrix
+
+    Parameters
+    -----------
+    ax_function : function
+        Takes in a guess vector and returns the Matrix-vector product.
+    preconditioner : function
+        Takes in a list of :py:class:`~psi4.core.Matrix` objects and a mask of active indices. Returns the preconditioned value.
+    guess : list of :py:class:`~psi4.core.Matrix`
+        Starting vectors, required
+    e_conv : float
+        Convergence tolerance for eigenvalues
+    r_conv : float
+        Convergence tolerance for residual vectors
+    no_eigs : int
+        Number of eigenvalues needed
+    maxiter : int
+        The maximum number of iterations this function will take.
+
+    Returns
+    -----------
+
+    Notes
+    -----------
+
+    Examples
+    -----------
+
+    """
+
+    print("Starting Davidson algo:")
+    print("options:")
+    for k,v in dict(e_conv=e_conv, r_conv=r_conv, no_eigs=no_eigs,
+            max_vecs_per_root = max_vecs_per_root, maxiter=maxiter).items():
+        if v == None:
+            v = 'None'
+        print("{:<25}{:.>100}".format(k,v))
```

Comment:
Can you have this function print out a header similar to the CG algorithm above? Also please use `core.print_out`.